pkg/disttask/framework/scheduler/balancer.go

```diff
@@ -98,6 +98,9 @@ func (b *balancer) balanceSubtasks(ctx context.Context, sch Scheduler, managedNo
 	if err != nil {
 		return err
 	}
+	if task.MaxNodeCount > 0 && len(eligibleNodes) > task.MaxNodeCount {
+		eligibleNodes = eligibleNodes[:task.MaxNodeCount]
+	}
```

Comment:
in `doBalanceSubtasks` we will filter nodes by whether it have enough nodes, better check node count after that